tripleawarclub.org/public_html/dice/dice.class.php

```diff
@@ -0,0 +1,287 @@
+	 */
+	function getEncryptionKey($date = null) {
+		// get old key
+		if($date) {
+			$dir = dirname(__FILE__);
+			$keyfile = fopen("$dir/keys/$date.dat", "r");
+		} else {
+			// get current key
+			if (!file_exists("key.dat")) {
+				$this->setEncryptionKey();
+			}
+			$keyfile = fopen("key.dat", "r");
+		}
+
+		if ($keyfile) {
+			$data = fread($keyfile, 8192);
+
+			$this->enc = unserialize($data);
+			fclose($keyfile);
+
+			return $this->enc;
+		} else {
+			exit("fatal error: Wrong date!");
+		}
+	}
+
+	function setEncryptionKey() {
+		$shouldWriteKey = true;
+		if (file_exists("key.dat")) {
+			$old = $this->getEncryptionKey();
+			$shouldWriteKey = copy("key.dat", "./keys/" . $old['date'] . ".dat");
+		}
+
+		if ($shouldWriteKey) {
+			$file = fopen("key.dat", "w");
+			$enc['key'] = $this->keygen();
+			$enc['date'] = $this->getDate();
+			$output = serialize($enc);
+			fputs($file, $output);
+			fclose($file);
+			chmod("key.dat", 0600);
+		}
+	}
+
+    function keygen() {
+		$tempstring = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
+		for($length = 1; $length < 24; $length++) {
+			$temp = str_shuffle($tempstring);
+			$char = mt_rand(0, strlen($temp));
+			$pass .= $temp[$char];
+		}
+		return $pass;
+	}
+
+	function encrypt_data($input) {
+		$this->getEncryptionKey();
+		$string = serialize($input);
+
+		$td = mcrypt_module_open('rijndael-256', '', 'cfb', '');
+    	$out['iv'] = mcrypt_create_iv (mcrypt_enc_get_iv_size($td), MCRYPT_RAND);
+    	mcrypt_generic_init($td, $this->enc['key'], $out['iv']);
+    	$encrypted_data = mcrypt_generic($td, $string);
+    	mcrypt_generic_deinit($td);
+    	mcrypt_module_close($td);
+
+		$out['data'] = rawurlencode(base64_encode($encrypted_data));
+		$out['iv'] = rawurlencode(base64_encode($out['iv']));
+
+		return $out;
+	}
+
+	function decrypt_data($date, $iv, $encrypted_data) {
+		$iv = base64_decode(rawurldecode($iv));
+		$encrypted_data = base64_decode(rawurldecode($encrypted_data));
+
+		if ($date != $this->getDate()) {
+			$encrypt_key = $this->getEncryptionKey($date);
+		} else {
+			$encrypt_key = $this->getEncryptionKey();
+		}
+
+		$td = mcrypt_module_open('rijndael-256', '', 'cfb', '');
+    	mcrypt_generic_init($td, $encrypt_key['key'], $iv);
+    	$decrypted_data = mdecrypt_generic($td, $encrypted_data);
+    	mcrypt_generic_deinit($td);
+    	mcrypt_module_close($td);
+
+		$output = unserialize($decrypted_data);
+
+		return $output;
+	}
+
+	function checkNewKeyNeeded() {
+		$now = $this->getDate();
+		$current_key = $this->getEncryptionKey();
+		if ($now != $current_key['date']) {
+			return true;
+		} else {
+			return false;
+		}
+	}
+
+	////////////////////////////////
+	//       dice and mail        //
+	////////////////////////////////
+	function createdice($numdice, $numsides) {
+		$i = 0;
+		while ($i <= $numdice-1) {
+			$dice[$i] = mt_rand(1,$numsides);
+			$i++;
+		}
```

Comment:
A for loop would be more appropiate here...
Also `$i <= $numdice-1` ? Are you for real? Why?!?!